### PR TITLE
Fix error for noncontiguous subset regions

### LINKED ISSUE #356: get_spectral_regions does not work when subset has multiple borders
I select a subset that is not continuous:
Screen Shot 2020-11-03 at 2 38 29 PM

then I run
`viz.get_spectral_regions()`

and I hit a bug

AttributeError                            Traceback (most recent call last)
<ipython-input-10-6d3b67f9b159> in <module>
----> 1 viz.get_spectral_regions()

~/miniconda3/envs/jdaviz1.0.3/lib/python3.8/site-packages/jdaviz/configs/specviz/helper.py in get_spectral_regions(self)
     37             as `specutils.SpectralRegion` objects.
     38         """
---> 39         regions = self.app.get_subsets_from_viewer("spectrum-viewer")
     40 
     41         spec_regs = {}

~/miniconda3/envs/jdaviz1.0.3/lib/python3.8/site-packages/jdaviz/app.py in get_subsets_from_viewer(self, viewer_reference, data_label)
    471                     #  the glue component objects
    472                     unit = value.data.coords.spectral_axis.unit
--> 473                     hi, lo = value.subset_state.hi, value.subset_state.lo
    474                     xcen = 0.5 * (lo + hi)
    475                     width = hi - lo

AttributeError: 'AndState' object has no attribute 'hi'

The call to `get_spectral_regions` works if the subset is a single selection with no holes.

### PR DESCRIPTION
**Pull request by javerbukh**
### Description

This pull request is to address the problem where running `viz.get_spectral_regions()` when there are noncontiguous subset regions fails. Regions within a subset are now returned and renamed "Subset 1_5306" where 5306 is the starting value of that particular region of the subset.

Fixes #356 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by rosteen** [comment]
It looks like your algorithm for finding the edges of the regions is slightly off - as you can see in the screenshots below, the first subregion is returned as the start of the subregion to the start of the next region, rather than the start of the subregion to the end of the subregion:

Screen Shot 2021-11-05 at 10 51 26 AM

Screen Shot 2021-11-05 at 10 51 40 AM

File: jdaviz/app.py [hunk 1]
```diff
@@ -542,6 +543,51 @@ def get_subsets_from_viewer(self, viewer_reference, data_label=None, subset_type
...
+                raise ValueError
+
+            current_edge = 0
+            combined_spec_region = None
+            for index in range(1, len(mask)):
+                # Find spot where mask == True is for a different region of the subset
+                # i.e. mask = [0, 1, 4, 5]
+                # mask[2] != mask[1] + 1
+                if mask[index] != mask[index - 1] + 1:
+                    subset_region = spec_axis_data[mask[current_edge]: mask[index] - 1]
```
Comment (rosteen) [inline jdaviz/app.py:573]: I suspect this is where a slight tweak is needed.

File: jdaviz/configs/specviz/tests/test_helper.py [hunk 2]
```diff
@@ -77,6 +78,66 @@ def test_get_spectra_label_redshift_warn(self):
...
+        self.spec_app.app.get_viewer("spectrum-viewer").apply_roi(XRangeROI(7300, 7800))
+
+        spec_region = self.spec_app.get_spectral_regions()
+
+        assert len(spec_region['Subset 1'].subregions) == 3
+        # Assert correct values for test with 3 subregions
+        assert_quantity_allclose(spec_region['Subset 1'].subregions[0][0].value,
+                                 6000., atol=1e-5)
+        assert_quantity_allclose(spec_region['Subset 1'].subregions[0][1].value,
+                                 6222.22222222, atol=1e-5)
```
Comment (rosteen) [inline jdaviz/configs/specviz/tests/test_helper.py:118]: Shouldn't the upper bound here match the ROI you defined? (6400, maybe slightly off depending on pixel spacing?)

**Review by rosteen — CHANGES_REQUESTED** [review]
Forgot to mark this as "Requested changes", see comments above.

File: jdaviz/configs/specviz/tests/test_helper.py (hunk 2 above)
Comment (javerbukh) [reply jdaviz/configs/specviz/tests/test_helper.py:118]: No it is exact, so 6444 will not be included.

File: jdaviz/app.py (hunk 1 above)
Comment (javerbukh) [reply jdaviz/app.py]: Good catch!!

**Comment by javerbukh** [comment]
@rosteen Good catch on the bug! Should be fixed now.

**Comment by rosteen** [comment]
I'm not sure if this is in scope or not, but this fails in a different way in Cubeviz, because that's handled by collapsing the data and using the `glue-astronomy` translator machinery rather than doing it manually for the 1D Specviz case. Traceback is below, although we might want to split it out into a separate issue/PR.

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/var/folders/ds/q5jd1xq53l9djq7lcc029d240002hz/T/ipykernel_18909/2076261966.py in <module>
----> 1 viz.app.get_subsets_from_viewer("spectrum-viewer")

~/projects/jdaviz/jdaviz/app.py in get_subsets_from_viewer(self, viewer_reference, data_label, subset_type)
    625                 # **Note** that the y values in this region object are not
    626                 #  useful, only the x values are.
--> 627                 region = subset_group.subsets[0].data.get_selection_definition(
    628                     format='astropy-regions')
    629                 regions[key] = region

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/glue/core/data.py in get_selection_definition(self, subset_id, format, **kwargs)
    371         handler = subset_state_translator.get_handler_for(format)
    372 
--> 373         return handler.to_object(subset, **kwargs)
    374 
    375 

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/glue_astronomy/translators/regions.py in to_object(self, subset)
    141             temp_sub2 = Subset(data=data)
    142             temp_sub2.subset_state = subset_state.state2
--> 143             return self.to_object(temp_sub1) | self.to_object(temp_sub2)
    144 
    145         elif isinstance(subset_state, XorState):

~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/glue_astronomy/translators/regions.py in to_object(self, subset)
    108                 return range_to_rect(data, 'y', subset_state.lo, subset_state.hi)
    109             else:
--> 110                 raise ValueError('Range subset state att should be either x or y pixel coordinate')
    111 
    112         elif isinstance(subset_state, MultiRangeSubsetState):

ValueError: Range subset state att should be either x or y pixel coordinate


```

**Comment by rosteen** [comment]
> @rosteen Good catch on the bug! Should be fixed now.

Just checked, the edge values look correct now.

**Comment by camipacifici** [comment]
Found the same problem as Ricky in Cubeviz, while it works just fine in Specviz and Mosviz.

File: jdaviz/configs/specviz/tests/test_helper.py (hunk 2 above)
Comment (rosteen) [reply jdaviz/configs/specviz/tests/test_helper.py:118]: For posterity's sake, the reason the values looked far off from what I expected is because the test spectrum only has 10 points (if I'm remembering correctly) and thus these are the values at the points nearest the specified bounds.

File: jdaviz/app.py [hunk 3]
```diff
@@ -542,6 +543,51 @@ def get_subsets_from_viewer(self, viewer_reference, data_label=None, subset_type
...
+                List of spectral axis values.
+            Returns
+            -------
+            combined_spec_region : `~specutils.SpectralRegion`
+                SpectralRegion object containing all subregions of the subset.
+            """
+            if len(mask) == 0:
+                # Mask should only be 0 if ApplyROI is used to incorrectly
+                # create subsets via the API
+                raise ValueError
```
Comment (rosteen) [inline jdaviz/app.py:564]:
```suggestion
                raise ValueError("Mask has length 0, ApplyROI may have been used incorrectly")
```

I dislike raising a bare/default error with no description.

**Review by rosteen — DISMISSED** [review]
This looks good to me other than the one suggestion I just left; if someone else would take a look and accept my suggestion/approve then we could get this merged. I except that Jesse won't be available to work on it again for a while, so if there are bigger changes needed I can take over. Note that the fix for the similar issue in Cubeviz is intended for a followup PR, since it will require `glue-jupyter` changes (I think) - see #959.

**Review by rosteen — COMMENTED** [review]
I actually found one more error, I should have a suggestion to fix it shortly.

File: jdaviz/app.py [hunk 4]
```diff
@@ -542,6 +543,51 @@ def get_subsets_from_viewer(self, viewer_reference, data_label=None, subset_type
...
+                        combined_spec_region += SpectralRegion(min(subset_region),
+                                                               max(subset_region))
+                    current_edge = index
+
+            # Get last region within the subset
+            if current_edge != index:
+                subset_region = spec_axis_data[mask[current_edge]: mask[index]]
+                # No if check here because len(mask) must be greater than 1
+                # so combined_spec_region will have been instantiated in the for loop
+                combined_spec_region += SpectralRegion(min(subset_region), max(subset_region))
```
Comment (rosteen) [inline jdaviz/app.py:587]: I managed to find a situation where this is attempting to add the region to None, trying to figure out how that was triggered in the code. Through the UI, I first made a Subset 1 with a few subregions, then created a Subset 2 with a single region. After that, I hit the `TypeError` that this says shouldn't be possible to hit.

Comment (rosteen) [reply jdaviz/app.py:587]: Ok, I found the cause here. If you make a single sub-region subset with the `replace` function, you get something different than if you do so with the `add` function. It looks the same on the front end, but Glue represents them differently and they end up looking different when returned by our `get_subsets_from_viewer` function. In the sample below, Subset 2 used `add` and Subset 3 used `replace`:

```
>>> s2.subset_state
<glue.core.subset.OrState at 0x7f9539e12970>

>>>s3.subset_state
<glue.core.subset.RangeSubsetState at 0x7f9548b09490>
```

And our app method returns:

```
'Subset 2': Spectral Region, 1 sub-regions:
   (5499.20556640625 Angstrom, 5897.93505859375 Angstrom) ,
 'Subset 3': <RectanglePixelRegion(center=PixCoord(x=7616.167200488471, y=0), width=418.5868159615429, height=0, angle=0.0 deg)>
```

It seems like we should return the same thing for a single-region Subset, no matter which method was used to create it, which I guess will require working around the Glue backend differences. This might be a good time to tag in @astrofrog and ask if something changed on the backend for Glue subsets recently, since as we've said before I'm 99% this had been working at some point in the past. (@astrofrog I know you're busy with astropy release right now, this is just a heads up for when you have time).

Comment (rosteen) [reply jdaviz/app.py:587]: Ah, wait, I realized why the model fitting used to work with subsets with subregions - we used to just grab e.g. `Subset 1` as the _data_, whereas now we've split out the data and spectral subsets into separate menus. So previously it was just grabbing the subset as a masked `Spectrum1D`, which worked fine for fitting. There were probably always problems with the subset retrieval for multiple subregions and we just didn't notice until we switched to trying to apply the subset separately for the model fitting.

**Review by duytnguyendtn — COMMENTED** [review]
Started my review in the midst of Ricky's investigation into the `replace`/'`add` subset discrepancy but I can at least chime in with my findings:

Tested this PR in Specviz in the Model Fitting plugin, and checked the API subset getters. In Specviz, I found that the subset `Subset 1` that I defined isn't showing up in the right area (i.e. the Spectral Region menu), but that isn't introduced by this PR. Just figured I should leave that breakcrumb. 
image
which then leads to the following traceback:
image

However, the actual fit looks good! The noncontiguious part worked! 
image
In addition, the API subset getters appear to show what I would expect from them: 
image

I agree with Ricky in https://github.com/spacetelescope/jdaviz/pull/951#discussion_r747760111, specifically:
> It seems like we should return the same thing for a single-region Subset, no matter which method was used to create it

I'll leave my specific thoughts in that thread. I won't approve for now until I get a chance to properly review the changes once we get that `replace`/`add` bug fixed

File: jdaviz/app.py (hunk 4 above)
Comment (duytnguyendtn) [reply jdaviz/app.py:587]: I would kind of expect any of our subset getters would resolve all regions into `specutils.spectra.SpectralRegion` objects at the end of the day, so I'm surprised we're getting a PixelRegion out of it.

**Comment by rosteen** [comment]
@duytnguyendtn @kecnry @ojustino @pllim This is now (again) ready for final review. It accomplishes two main things:

1. `get_subsets_from_viewer` now successfully returns `SpectralRegion`s for all 1D spectral regions, including ones that include multiple subregions.
2. Fitting a spectrum while applying a subset that includes multiple regions works again in Specviz (this broke when we split the Data and Subset selections into separate dropdown menus). 

Note that both of these will require more work to get them working in Cubeviz, but that is out of scope for this PR. Also, as a side benefit I got to delete some code in the Specviz helper due to (1). 

I included a quick example below showing different model fits to the whole spectrum (purple) vs. with a multiregion subset applied (green) but...test it yourself too 😁 .

Screen Shot 2021-11-12 at 2 39 18 PM

File: jdaviz/app.py [hunk 5]
```diff
@@ -545,15 +545,69 @@ def get_subsets_from_viewer(self, viewer_reference, data_label=None, subset_type
...
+                # create subsets via the API
+                raise ValueError("Mask has length 0, ApplyROI may have been used incorrectly")
+
+            current_edge = 0
+            combined_spec_region = None
+            for index in range(1, len(mask)):
+                # Find spot where mask == True is for a different region of the subset
+                # i.e. mask = [0, 1, 4, 5]
+                # mask[2] != mask[1] + 1
+                if mask[index] != mask[index - 1] + 1:
```
Comment (pllim) [inline jdaviz/app.py:574]: This assumes that `mask` is a well-behaved list (no repeats, monotonically increasing)? What if someone inputs crazy values like `mask = [-1, 14, -10, 0]`? Is this a use case we need to worry about?

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py [hunk 6]
```diff
@@ -255,23 +259,20 @@ def vue_data_selected(self, event):
...
-        for key, region in temp_subsets.items():
-            if type(region) == RectanglePixelRegion:
-                temp_dict[key] = region
-        self._spectral_subsets = temp_dict
-        self.spectral_subset_items = ["Entire Spectrum"] + sorted(temp_dict.keys())
+        self._spectral_subsets = self.app.get_subsets_from_viewer("spectrum-viewer",
+                                                                  subset_type="spectral")
+        self.spectral_subset_items = ["Entire Spectrum"] + sorted(self._spectral_subsets.keys())
 
+    '''
```
Comment (pllim) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:275]: Is this also pending cubeviz fix or is it dead code now?

File: jdaviz/app.py (hunk 5 above)
Comment (rosteen) [reply jdaviz/app.py:574]: The logic there is ok, because `mask` there is actually the indices where the mask derived from the subset on the spectrum is `True` - the variable name is a little confusing. The code fed into that `mask` argument is `np.where(value.to_mask() == True)`, which should work fine based on the `Subset.to_mask()` logic, I think.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py (hunk 6 above)
Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:275]: Good catch, I should have uncommented this for Cubeviz.

Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: Fixed, Cubeviz is happy (well, as happy as it was) again.

**Comment by duytnguyendtn** [comment]
> `get_subsets_from_viewer` now successfully/consistently returns `SpectralRegions` for all 1D spectral regions, including ones that include multiple subregions.

Testing the [merge commit based on 4adc8f6](https://github.com/spacetelescope/jdaviz/commit/e17b4867671183b83bfb5a17c5ab6685a6fa993c), I'm still seeing `RectangularPixelRegion` rather than `SpectalRegion`. Just a simple `replace` region defined. Did I do something wrong?

image

**Comment by rosteen** [comment]
Yes, you're in Cubeviz instead of Specviz.

**Comment by duytnguyendtn** [comment]
Ahh, I thought this was fixed in all the viz tools, not just specviz. Checking now...

**Comment by rosteen** [comment]
There are additional complications in Cubeviz, because of the need to collapse to the 1D spectrum it handles things differently.

**Review by duytnguyendtn — APPROVED** [review]
Looks good to me! Thanks Ricky for taking this over and fixing the bug! 
image

**Review by camipacifici — APPROVED** [review]
Works well for me!

**Comment by rosteen** [comment]
Dev test failure is unrelated, merging.
